device/devicegate/filterHandler.go

```diff
@@ -0,0 +1,88 @@
+package devicegate
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/go-kit/kit/log/level"
+	"github.com/xmidt-org/webpa-common/logging"
+	"github.com/xmidt-org/webpa-common/xhttp"
+)
+
+// FilterHandler is an http.Handler that can get, add, and delete filters from a devicegate Interface
+type FilterHandler struct {
+	Gate Interface
+}
+
+func (fh *FilterHandler) ServeHTTP(response http.ResponseWriter, request *http.Request) {
+	logger := logging.GetLogger(request.Context())
+
+	method := request.Method
+	if method == http.MethodGet {
+		response.Header().Set("Content-Type", "application/json")
+		JSON, _ := json.Marshal(fh.Gate)
+		fmt.Fprintf(response, `%s`, JSON)
+	} else if method == http.MethodPost || method == http.MethodPut || method == http.MethodDelete {
+		var message FilterRequest
+		msgBytes, err := ioutil.ReadAll(request.Body)
+		request.Body.Close()
+
+		if err != nil {
+			logger.Log(level.Key(), level.ErrorValue(), logging.MessageKey(), "could not read request body", logging.ErrorKey(), err)
+			xhttp.WriteError(response, http.StatusBadRequest, err)
+			return
+		}
+
+		if err := json.Unmarshal(msgBytes, &message); err != nil {
+			logger.Log(level.Key(), level.ErrorValue(), logging.MessageKey(), "could not unmarshal request body", logging.ErrorKey(), err)
+			xhttp.WriteError(response, http.StatusBadRequest, err)
+			return
+		}
+
+		if allow, err := checkRequestDetails(method, message, fh.Gate); !allow {
+			logger.Log(level.Key(), level.ErrorValue(), logging.MessageKey(), err)
+			xhttp.WriteError(response, http.StatusBadRequest, err)
+			return
+		}
+
+		if method == http.MethodPost || method == http.MethodPut {
+			if _, created := fh.Gate.SetFilter(message.Key, message.Values); created {
+				response.WriteHeader(http.StatusCreated)
+			} else {
+				response.WriteHeader(http.StatusOK)
+			}
+		} else if method == http.MethodDelete {
+			fh.Gate.DeleteFilter(message.Key)
+			response.WriteHeader(http.StatusOK)
+		}
+
+		filtersJSON, _ := json.Marshal(fh.Gate)
+		logger.Log(level.Key(), level.InfoValue(), logging.MessageKey(), "gate filters updated", "filters", string(filtersJSON))
```

Comment:
General purpose logging like this makes better sense as a decorator.  That makes this code less noisy and avoids repeating yourself.  For example:

```
type GateLogger struct {
    Logger log.Logger
}

func (gl GateLogger) Then(next http.handler) http.Handler {
    return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {
        // first, delegate to the next handler to do whatever is necessary for the gate
        next.ServeHTTP(response, request)

        // now, perform the logging.  the "filtersJSON" can be communicated through the request's context
        // it's probably easiest to set the entire Gate object in the context inside the next handler or another decorator

        if filtersJSON, err := json.Marshal(/* get the Gate from request.Context() */); err == nil {
            logger.Log(level.Key(), level.InfoValue(), logging.MessageKey(), "gate filters updated", "filters", string(filtersJSON))
        } else {
            // NEVER SWALLOW ERRORS! log the error here
            // sometimes, it is appropriate to panic since this indicates that something's buggy with the gate
        }
    })
}

// You can use MiddlewareFunc to decorate things
gl := GateLogger{
    Logger: /* logger goes here! */
}

r := mux.Router()
r.MiddlewareFunc(gl.Then) // this will apply the middleware to any routes defined afterward

// You can also use alice to decorate each handler separately
c := alice.NewChain(gl.Then)
r.Path("/foo").Handle(c.Then(myHandler))
r.Path("/bar").Handle(nonGateHandler)
```